controller/src/controller/models/domain/rspimage.py

```diff
@@ -549,3 +553,59 @@ def _resolve_duplicate(self, new: RSPImage, old: RSPImage) -> None:
                     other.aliases.add(new.tag)
             old.aliases.add(new.tag)
             self._by_digest[new.digest] = new
+
+    def filter(
+        self, policy: RSPImageFilterPolicy, age_basis: datetime
+    ) -> Self:
+        """Apply a filter policy and return the remaining tags.
+
+        Parameters
+        ----------
+        policy
+            Policy governing tag filtering.
+        age_basis
+            Timestamp to use as basis for image age calculation.
+
+        Returns
+        -------
+        RSPImageTagCollection
+            Tags remaining after policy application.
+        """
+        images: list[RSPImage] = []
+        for category in RSPImageType:
+            images.extend(
+                self._apply_category_policy(policy, category, age_basis)
+            )
+        return type(self)(images)
+
+    def _apply_category_policy(
+        self,
+        policy: RSPImageFilterPolicy,
+        category: RSPImageType,
+        age_basis: datetime,
+    ) -> list[RSPImage]:
+        candidates = list(self._by_type[category])
+        remainder: list[RSPImage] = []
+        cat_policy = policy.policy_for_category(category)
+        if cat_policy is None:
+            remainder.extend(candidates)
+            return remainder
+        for image in candidates:
+            if cat_policy.number is not None and cat_policy.number <= len(
+                remainder
+            ):
+                break
+            if image.date is not None and cat_policy.age is not None:
+                cutoff_date = age_basis - cat_policy.age
+                if image.date < cutoff_date:
+                    continue
+            if (
+                image.version is not None
+                and cat_policy.cutoff_version is not None
+            ):
+                img_ver = semver.Version.parse(image.version)
+                cat_pol_ver = semver.Version.parse(cat_policy.cutoff_version)
+                if img_ver < cat_pol_ver:
+                    continue
```

Comment:
I'd still like to see the threshold calculations factored out of the loop rather than recalculating the cutoff date on every time through the loop. I don't care about the performance hit, but I think it would make the code easier to read.

Reply:
I don't, but I'm willing to change it.